fe/estimator.py

```diff
@@ -1,170 +1,28 @@
-
-def simulate(lamb, box, x0, v0, final_potentials, integrator, equil_steps, prod_steps,
-    x_interval=1000, du_dl_interval=5):
-    """
-    Run a simulation and collect relevant statistics for this simulation.
-
-    Parameters
-    ----------
-    lamb: float
-        lambda parameter
-
-    box: np.array
-        3x3 numpy array of the box, dtype should be np.float64
-
-    x0: np.array
-        Nx3 numpy array of the coordinates
-
-    v0: np.array
-        Nx3 numpy array of the velocities
-
-    final_potentials: list
-        list of unbound potentials
-
-    integrator: timemachine.Integrator
-        integrator to be used for dynamics
-
-    equil_steps: int
-        number of equilibration steps
-
-    prod_steps: int
-        number of production steps
-
-    x_interval: int
-        how often we store coordinates. if x_interval == 0 then
-        no frames are returned.
-
-    du_dl_interval: int
-        how often we store du_dls. if du_dl_interval == 0 then
-        no du_dls are returned
-
-    Returns
-    -------
-    SimulationResult
-        Results of the simulation.
-
-    """
-    all_impls = []
-    bonded_impls = []
-    nonbonded_impls = []
-
-    # set up observables for du_dps here as well.
-    du_dp_obs = []
-
-    for bp in final_potentials:
-        impl = bp.bound_impl(np.float32)
-        if isinstance(bp, potentials.Nonbonded):
-            nonbonded_impls.append(impl)
-        else:
-            bonded_impls.append(impl)
-        all_impls.append(impl)
-        du_dp_obs.append(custom_ops.AvgPartialUPartialParam(impl, 5))
-
-    if integrator.seed == 0:
-        integrator = copy.deepcopy(integrator)
-        integrator.seed = np.random.randint(np.iinfo(np.int32).max)
-
-    intg_impl = integrator.impl()
-    # context components: positions, velocities, box, integrator, energy fxns
-    ctxt = custom_ops.Context(
-        x0,
-        v0,
-        box,
-        intg_impl,
-        all_impls
-    )
-
-    # equilibration
-    equil_schedule = np.ones(equil_steps)*lamb
-    ctxt.multiple_steps(equil_schedule)
-
-    for obs in du_dp_obs:
-        ctxt.add_observable(obs)
-
-    prod_schedule = np.ones(prod_steps)*lamb
-
-    full_du_dls, xs = ctxt.multiple_steps(prod_schedule, du_dl_interval, x_interval)
-
-    # keep the structure of grads the same as that of final_potentials so we can properly
-    # form their vjps.
-    grads = []
-    for obs in du_dp_obs:
-        grads.append(obs.avg_du_dp())
-
-    result = SimulationResult(xs=xs, du_dls=full_du_dls, du_dps=grads)
-    return result
-
-
 FreeEnergyModel = namedtuple(
     "FreeEnergyModel",
-    ["unbound_potentials", "client", "box", "x0", "v0", "integrator", "lambda_schedule", "equil_steps", "prod_steps"]
+    ["dGs", "dG_grads"]
 )
 
 gradient = List[Any] # TODO: make this more descriptive of dG_grad structure
 
 def _deltaG(model, sys_params) -> Tuple[Tuple[float, List], np.array]:
```

Comment:
This is needed to properly compute the gradient, otherwise backpropagation would break. deltaG is still a function that takes in `sys_params` and returns a `dG`, and is just `dG/dsys_params`. If `sys_params` were not an argument, then the gradient isn't even defined!